### DISCUSSION TITLE
Is there browsers built with .NET

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/runtime#65276] [General]
Chromium is built using c, would be nice to see .NET able to build a browser, it would be a living advertisement

**Comment by John0King** [comment]
@Xyncgas I'm afraid the answer is no,  there are too many things that needed by a browser, for examples:
-  encryption (need OpenSSL probably)
- OS layer method
- UI drawing(DirextX, OpenGL) Video Decoding, Image Decoding
- WebGL implementation, 
- javascript engine.
- web assembly engine
- and so on

and each of them need  p/invoke with C/C++ or use it's native stack,    unless C# can direct reference the C/C++ header file, I think it will not for future neither.

**Reply by Clockwork-Muse** [reply]
... There's nothing stopping any of those being written in C#.  
It's particularly egregious that you mention encryption on that list, because dotnet **has** an encryption library in the runtime, and will transparently handle SSL when you make HTTP requests.  (Granted, it's wrapping OpenSSL on linux, and OS libraries on Windows, but this isn't because "not possible", but rather "the hard work is already done").

> OS layer method

It's unclear what you mean by this, but **no** language just "natively" interacts with an OS.  They _all_ have some sort of runtime that needs to be compiled and called, including C.  Most other languages bind and call to C APIs on the OS because it's usually already there, because the OS was implemented in C/C++, but that is unlikely to be the case if the OS was implemented by something else.

> UI drawing(DirextX, OpenGL) Video Decoding, Image Decoding
> WebGL implementation,

Various wrapper libraries exist, mostly for historical or performance reasons, but nothing prevents any of those from being implemented in C# itself.

> javascript engine.

[Ecma script interpreter](https://github.com/sebastienros/jint) (not an endorsement).  Missing features, but other projects exist, and alternate implementation styles.

> web assembly engine

[wasm interpreter](https://github.com/jonathanvdc/cs-wasm) (not an endorsement).

**Reply by Xyncgas** [reply]
We can always start with abstract classes

**Reply by Clockwork-Muse** [reply]
.... what?  That's not relevant

**Reply by Xyncgas** [reply]
You've made some pretty good points

**Reply by John0King** [reply]
A browse build using .net is possible but it's hard or harder  because you need to create your own `extern` method to calling into those native, and C/C++ just got much easier , that's why i said "unless C# can direct reference the C/C++ header file, I think it will not for future neither" ,  and probably can  in  future because `Source Generator`.

**Reply by Clockwork-Muse** [reply]
Even C/C++ have "extern" calls, essentially (although depending on a few factors the compiler might handle some of it for you).  That aside, pretty much all of the work of creating the C# external calls has already been done by the other libraries, so at that point the only thing required would be to stitch the C# libraries together.

**Comment by Clockwork-Muse** [comment]
To some limited extent, it looks like it has [some form of limited browser already](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.webbrowser?view=windowsdesktop-6.0), at least on Windows.

That aside, I'm not sure that this would markedly increase the appeal of the language.  Most - nearly all, besides a _vanishingly_ small percentage - people that use a browser have no idea how to program.  If they do, they are unlikely to change their perception of the language just based on a browser.

**Comment by rogihee** [comment]
I heard rumors of DuckDuckGo building a WPF based browser...

**Reply by Xyncgas** [reply]
oh really, is there links

**Comment by huoyaoyuan** [comment]
Browser is getting more and more complex these days. EdgeHTML, Trident(IE) and Presto(Opera) are discontinued. Firefox is not going well.

**Reply by Xyncgas** [reply]
I am more than optimistic, a browser built with .NET is a platform for starter, at least it's another application for people to use, that alone enables a lot due to the huge eco system.

Popularity alone shouldn't become a reason to stop something from emerging
